src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1378,43 +1363,64 @@ private static bool HasIdentityConversionInternal(TypeSymbol type1, TypeSymbol t
+            var compareKind = TypeCompareKind.AllIgnoreOptions |
+                (includeNullability ?
+                    TypeCompareKind.CompareNullableModifiersForReferenceTypes |
+                        (unknownNullableModifierMatchesAny ? TypeCompareKind.UnknownNullableModifierMatchesAny : 0) :
+                    0);
             return type1.Equals(type2, compareKind);
         }
 
         private bool HasIdentityConversionInternal(TypeSymbol type1, TypeSymbol type2)
         {
-            return HasIdentityConversionInternal(type1, type2, IncludeNullability);
+            return HasIdentityConversionInternal(type1, type2, IncludeNullability, UnknownNullableModifierMatchesAny);
         }
 
-        internal static bool HasTopLevelNullabilityIdentityConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
+        internal bool HasTopLevelNullabilityIdentityConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
         {
+            if (!IncludeNullability)
+            {
+                return true;
+            }
             bool? sourceIsNullable = source.IsNullable;
             bool? destinationIsNullable = destination.IsNullable;
-            return sourceIsNullable == null || destinationIsNullable == null || sourceIsNullable == destinationIsNullable;
-        }
-
-        internal static bool HasTopLevelNullabilityImplicitConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
-        {
-            return HasTopLevelNullabilityImplicitConversion(source.IsNullable, destination.IsNullable);
+            if (sourceIsNullable == destinationIsNullable)
+            {
+                return true;
+            }
+            return UnknownNullableModifierMatchesAny &&
+                (sourceIsNullable == null || destinationIsNullable == null);
         }
 
-        internal static bool HasTopLevelNullabilityImplicitConversion(bool? sourceIsNullable, bool? destinationIsNullable)
+        internal bool HasTopLevelNullabilityImplicitConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
         {
-            return sourceIsNullable != true || destinationIsNullable != false;
+            if (!IncludeNullability)
+            {
+                return true;
+            }
+            bool? sourceIsNullable = source.IsNullable;
+            bool? destinationIsNullable = destination.IsNullable;
+            if (sourceIsNullable == destinationIsNullable)
+            {
+                return true;
+            }
+            switch (sourceIsNullable)
+            {
+                case true:
+                    return UnknownNullableModifierMatchesAny && destinationIsNullable == null;
+                case false:
+                    return true;
+                default:
+                    return UnknownNullableModifierMatchesAny || destinationIsNullable == true;
+            }
```

Comment:
>} [](start = 12, length = 1)

I'd be very tempted to replace this method with two lookups into `bool[3][3]` arrays. I think it would be more readable. (I personally can't verify the current implementation without enumerating the cases on a piece of paper...)

Reply:
I think that would require two [3][3] arrays though, to include the value of `UnknownNullableModiferMatchesAny`.

---
In reply to: [221332482](https://github.com/dotnet/roslyn/pull/29727#discussion_r221332482) [](ancestors = 221332482)

Reply:
Yes, I meant two such arrays.

---
In reply to: [221369758](https://github.com/dotnet/roslyn/pull/29727#discussion_r221369758) [](ancestors = 221369758,221332482)

Reply:
How about something like this:
```C#
if (UnknownMatchesAny && (source == null || destination == null))
{
    return true;
}
// ? <- ~ <- !
switch (source)
{
   case true:
        return destination == true;
   default:
        return ...
   case false;
        return true;
}
```

---
In reply to: [221397045](https://github.com/dotnet/roslyn/pull/29727#discussion_r221397045) [](ancestors = 221397045,221369758,221332482)